### PR TITLE
Prevent Dynamo graph fragmentation in GPTNeoX with torch.baddbmm fix

### LINKED ISSUE #24940: TorchDynamo graph needlessly fragmented for GPTNeoX due to baddbmm type mistake
### System Info

- `transformers` version: 4.32.0.dev0
- Platform: Linux-5.19.0-42-generic-x86_64-with-glibc2.27
- Python version: 3.10.10
- Huggingface_hub version: 0.15.1
- Safetensors version: 0.3.1
- Accelerate version: 0.21.0
- Accelerate config:    not found
- PyTorch version (GPU?): 2.0.1+cu117 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: No
- Using distributed or parallel set-up in script?: No

### Who can help?

@ArthurZucker @younesbelkada 

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

```py
from transformers import AutoModelForCausalLM
import torch


def debug_backend(gm: torch.fx.GraphModule, example_inputs: list[torch.Tensor]):
    print("debug_backend() called with FX graph:")
    gm.graph.print_tabular()
    return gm.forward  # return a python callable

model = AutoModelForCausalLM.from_pretrained("EleutherAI/pythia-160m")
jitted = torch.compile(model, backend=debug_backend)

jitted(**model.dummy_inputs)
```
The output is too long to fit in a comment, so you'll have to run the code yourself. It features `"debug_backend() called with FX graph:"` being printed several times, each time followed with a fragment of the whole computation graph. This is not expected since NeoX has no data-dependent control flow.

### Expected behavior

The `torch.compile` backend should only be called once, and therefore `"debug_backend() called with FX graph:"` should only appear once, because GPT NeoX does not actually require any data-dependent control flow.

I've already checked that this can be fixed by turning `GPTNeoXAttention.norm_factor` into a Python scalar instead of a tensor. This is actually what `torch.baddbmm` expects for its `alpha` parameter; it's [supposed to be](https://pytorch.org/docs/stable/generated/torch.baddbmm.html) a scalar. But it seems to silently convert tensors into scalars, so this doesn't cause a crash in normal use.
Captura de pantalla 2023-07-19 a la(s) 5 27 42 p m

The exact fix is, in `modeling_gpt_neox.py`, replace lines 103-107 with:
```py
self.norm_factor = self.head_size ** -0.5
```
and replace the `baddbmm` call inside `_attn` with:
```py
attn_scores = torch.baddbmm(
    attn_scores,
    query,
    key.transpose(1, 2),
    beta=1.0,
    alpha=self.norm_factor,
)
```

### PR DESCRIPTION
**Pull request by norabelrose**
# What does this PR do?

Fixes #24940.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@ArthurZucker and @younesbelkada

### CODE REVIEW

**Comment by norabelrose** [comment]
I'm struggling to see what's wrong with my code style since when I run `black` locally it says it's fine.

**Review by ArthurZucker — COMMENTED** [review]
Hey! Thanks for submitting a PR! You should be using `make style` to properly format the code. 

My only concern here is that the dtype is not set. A small test like this shows that this will  have some effect for some head size if you are doing `float16` vs `float32` for example:
```python 
for i in range(1,4096):
  print(f"{(i ** -0.5) == torch.tensor(1, dtype = torch.float16)/ torch.sqrt(torch.tensor(i))}")
```

Thus I am a bit reluctant to change this currently.

**Comment by norabelrose** [comment]
> Hey! Thanks for submitting a PR! You should be using `make style` to properly format the code.

This doesn't seem to work for me. `make style` makes a ton of changes to _other_ files that I didn't modify, but doesn't change my code at all.

**Comment by norabelrose** [comment]
> My only concern here is that the dtype is not set. A small test like this shows that this will have some effect for some head size

I suppose I can manually cast `norm_factor` to the appropriate dtype and then convert back to a Python scalar. On the other hand, inverse square root isn't going to have _any_ numerical error as long as the head size is a power of 2, which it almost always is (in fact, it's almost always 64). So in practice this shouldn't matter at all.

Also, presumably the ground truth should be the Eleuther `gpt-neox` implementation. That implementation seems to already encode `norm_factor` as a Python scalar and uses `math.sqrt`, see [here](https://github.com/EleutherAI/gpt-neox/blob/408e29d9c746a02d842917bb7447c5c4be0b42d4/megatron/model/transformer.py#L298). It passes a Python scalar to `torch.baddbmm` [here](https://github.com/EleutherAI/gpt-neox/blob/408e29d9c746a02d842917bb7447c5c4be0b42d4/megatron/model/transformer.py#L420).

So in fact, our current implementation is already "wrong" with respect to the original, and this PR would correct the discrepancy. Although, as I said, in almost all cases this will make no difference.

**Comment by ArthurZucker** [comment]
Ok! Got what you mean. Part of the code was added in #22888 in order to have proper float16 casting. Yes we should match the original results (which is currently the case) but we have also other functionalities on which a lot of users rely! 
Let's make sure that the value is casted to the correct type for the next computations. 

(make style is probably behaving wrongly because of the black / ruff versioning)

**Comment by ArthurZucker** [comment]
Sorry but no, in transformers we autocast the weights to the required dtype. You can try the following:
```python 
>>> from transformers import GPTNeoXModel
>>> import torch 

>>> model = GPTNeoXModel.from_pretrained("EleutherAI/gpt-neox-20b", torch_dtype = torch.float16)
>>> print(model.layers[1].attention.norm_factor)
tensor(9.7969, dtype=torch.float16)
```
So the dtype was properly handle. We initialized it with float32 but then it can be casted to any dtype. 

For the red ci run `make style`!

**Comment by norabelrose** [comment]
> Sorry but no, in transformers we autocast the weights to the required dtype.

But this isn't a _weight_, though. It's a constant value deterministically computed from the config.

> For the red ci run `make style`!

I'm sorry, but this simply does not work! It changes over 200 files that I haven't changed, but doesn't do anything to the files I did change. Please help me here, can you just clone my fork and somehow fix it or figure out what's wrong?

**Comment by ArthurZucker** [comment]
Just pushed the changes! 
You are right it is not a weight, but it was 😓 
Which mean that this fix would be breaking previous outputs. Since the model is pretty old, would like to just cast the norm factor to the correct dtype (since I am not sure we can be 100% sure all previous outputs will be the sameWDYT?)

**Comment by norabelrose** [comment]
I'm not sure what you mean by "cast the norm factor to the correct dtype." At the very least, `norm_factor` _must_ be a Python scalar in order to fix the Dynamo graph fragmentation issue, which means that it has to be a double when we call `torch.baddbmm`.

If you're saying we cast it to the model's dtype and then cast it back— I _suppose_ we could do that, but I would rather not. Right now there is _technically_ a discrepancy between our implementation and the original NeoX implementation. If we merge this PR as-is, the discrepancy would be fixed.

That said, this discrepancy will virtually never matter since `head_size` is always a power of 2 in all the pretrained checkpoints. They will behave identically.

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py [hunk 1]
```diff
@@ -258,7 +254,7 @@ def _attn(self, query, key, value, attention_mask=None, head_mask=None):
             query,
             key.transpose(1, 2),
             beta=1.0,
-            alpha=(torch.tensor(1.0, dtype=self.norm_factor.dtype, device=self.norm_factor.device) / self.norm_factor),
+            alpha=self.norm_factor,
```
Comment (ArthurZucker) [inline src/transformers/models/gpt_neox/modeling_gpt_neox.py:257]: You are right! 
Sorry for dragging this one, TLDR: 
- the `self.head_size ` is always `self.hidden_size / self.num_attention_heads`. But there's  a check to make sure it is divisible! (So basically it's an int)
- the only way for this to break something is if  `self.head_dim = 65504`. In this case, the previous behaviour had: 
(I know it's a huge head dim 🤣 ) 
```python 
>>> torch.sqrt(torch.tensor(65504, device = "cuda", dtype = torch.float32)).to(torch.float16)
tensor(255.8750, device='cuda:0', dtype=torch.float16)
>>> # vs without casting
>>> torch.sqrt(torch.tensor(65504, device = "cuda", dtype = torch.float32))
tensor(255.9375, device='cuda:0')
```
Should be okay to have this

**Review by ArthurZucker — APPROVED** [review]

**Review by sgugger — APPROVED** [review]

**Comment by ArthurZucker** [comment]
Thanks @norabelrose for bearing with me and merging this!
